core/src/main/java/com/novoda/merlin/service/Ping.java

```diff
@@ -7,24 +7,24 @@
 
     private final String hostAddress;
     private final HostPinger.ResponseCodeFetcher responseCodeFetcher;
+    private final ResponseCodeValidator validator;
 
-    Ping(String hostAddress, HostPinger.ResponseCodeFetcher responseCodeFetcher) {
+    Ping(String hostAddress, HostPinger.ResponseCodeFetcher responseCodeFetcher, ResponseCodeValidator validator) {
         this.hostAddress = hostAddress;
         this.responseCodeFetcher = responseCodeFetcher;
+        this.validator = validator;
     }
 
     public boolean doSynchronousPing() {
         MerlinLog.d("Pinging : " + hostAddress);
         try {
-            responseCodeFetcher.from(hostAddress);
+            return validator.isResponseCodeValid(responseCodeFetcher.from(hostAddress));
         } catch (RequestException e) {
-            if (e.causedByIO()) {
-                return false;
+            if (!e.causedByIO()) {
+                MerlinLog.e("Ping task failed due to " + e.getMessage());
```

Comment:
All we were doing previously was throwing it one level up to the AsyncTask implementation... which then did this.